### ISSUE TITLE
spirv-fuzz: Make a synonym via OpSelect

### ISSUE THREAD

**Issue by paulthomson** [KhronosGroup/SPIRV-Tools#3776]

**Comment by afd** [comment]
We have TransformationAddSynonym, which makes synonyms via various opcodes, and it would be good to add OpSelect as one of the possibilities.

**Comment by afd** [comment]
@Vasniktel I cannot decide whether to try to add this as a new case in TransformationAddSynonym, or make it a separate transformation.

Three things make it not fit the pattern of the other cases in TransformationAddSynonym:

1. We want to turn %r into either "OpSelect %true %r %anything" or "OpSelect %false %r %anything".  If we want both of these cases we would need to encode whether "true" or "false" is desired, or have another transformation to invert an OpSelect.

2. We would need to supply the id for %anything - the other cases don't have this.

3. When operating on vectors, OpSelect can either use a scalar boolean condition or a vector condition.  The vector condition case could be quite interesting.  For two vectors %v and %w we could end up with a vector %u such that some components of %u are synonyms for some components of %v and some components of %w.

Perhaps this merits its own transformation and fuzzer pass; what do you think?

Or: the OpSelect case with a scalar condition could be folded in to TransformationAddSynonym (it's very similar to OpCopyObject), and another transformation could be added for the vector case.

**Comment by afd** [comment]
@Vasniktel Maybe we should revisit this as a good first issue?

**Comment by Vasniktel** [comment]
Yes, I think so. Also, regarding your previous comment, I think we might be able to use `TransformationAddSynonym`:
> We want to turn %r into either "OpSelect %true %r %anything" or "OpSelect %false %r %anything". If we want both of these cases we would need to encode whether "true" or "false" is desired, or have another transformation to invert an OpSelect.

We can use `%true` by default and have a separate transformation to swap the operands (something similar to what `TransformationSwapConditionalBranchOperands` does). 

> We would need to supply the id for %anything - the other cases don't have this.

We can use an irrelevant 0 constant by default for `%anything` - we already have a transformation that would replace it with anything.

> When operating on vectors, OpSelect can either use a scalar boolean condition or a vector condition. The vector condition case could be quite interesting. For two vectors %v and %w we could end up with a vector %u such that some components of %u are synonyms for some components of %v and some components of %w.

We could just use a simple case in the `TransformationAddSynonym` when we select between either all components of `%v` or all components of `%w`. We could then extend the operand-swapping transformation from above to support intermixing vector components.

The advantage of the approach above is that the operand-swapping transformation would support all `OpSelect` instructions as opposed to only those, that add synonyms.

That being said, I don't have a particularly strong opinion on this. Having a separate transformation for `OpSelect` synonyms is fine with me too.

**Comment by Vasniktel** [comment]
I think it's better to focus only on scalar operands right now. The original and transformed modules will look as follows:
- original:
```
...
%a = ... ; some instruction of scalar type - say %type
...
```
- transformed:
```
...
%a = ... ; the same instruction
%b = OpSelect %type %false %irrelevant_zero %a
...
```

In the above:
- `%type` is some scalar type.
- `%false` is a non-irrelevant boolean constant (`%false = OpConstantFalse %bool`)
- `%irrelevant_zero` is an irrelevant scalar zero constant with type `%type`.

Also, now the fact manager has the fact that `%a` and `%b` are synonymous.

Later, we can extend this to support vector operands and maybe swapping operands of the `OpSelect` but that should be a different PR.
